_posts/2023-02-28-kani-internship-projects-2022-stubbing.md

```diff
@@ -0,0 +1,361 @@
+    let n = n_64 as usize;
+    if n < FACT.len() {
+        Some(FACT[n])
+    } else {
+        None
+    }
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(factorial, stub_factorial)]
+fn verify_factorial() {
+    let n = kani::any();
+    kani::assume(n > 0);
+    if let Some(value) = factorial(n) {
+        assert_eq!(value, n * factorial(n - 1).unwrap());
+    }
+}
+```
+
+Let's run this harness with Kani!
+
+```bash
+cargo kani --enable-unstable --enable-stubbing --harness verify_factorial
+```
+
+```
+SUMMARY:
+ ** 0 of 7 failed
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.22135049s
+```
+
+Now Kani verifies the example successfully, and in less than 1 second!
+This time, Kani is able to complete the verification because our stub avoids recursion altogether.
+
+## Risks of Stubbing
+
+Stubbing is a feature that comes with great power and, as such, it should be used with caution.
+It's the developer's responsibility to **ensure that a stub replacing another function soundly models its behavior**.
+This is normally not the case in testing, where developers writing *mock-ups* tend to setup a concrete version of an object (an *under-approximation* of the model).
+
+For example, let's suppose you're attempting to stub a call to [`serde_json::from_slice`](https://docs.rs/serde_json/latest/serde_json/de/fn.from_slice.html) in your harness.
+This isn't a strange situation since deserialization often leads to bad performance.
+
+```rust
+#[derive(Deserialize)]
+#[cfg_attr(kani, derive(kani::Arbitrary))]
+pub struct MyStruct {
+    // -- snip -- 
+}
+
+#[cfg(kani)]
+fn stub_deserialize<S, T>(_data: &[u8]) -> serde_json::Result<T>
+where
+    T: kani::Arbitrary,
+{
+    Ok(kani::any())
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(serde_json::from_slice, stub_deserialize)]
+fn verify_with_deserialization() {
+    let data = symbolic_slice();
+    let my_struct = serde_json::from_slice::<MyStruct>(&data);
+    // -- snip -- 
+}
+```
+
+Here, we're using `kani::any()` to generate a symbolic `MyStruct`.
+This is easy to do since we're deriving `kani::Arbitrary` for `MyStruct`.
+
+In principle, this may look like an **over-approximation** because we generate a symbolic value for `MyStruct`.
+But actually, we're missing something:
+[`serde_json::Result<T>`](https://docs.rs/serde_json/latest/serde_json/type.Result.html) can also return a `serde_json::Error` if the deserialization fails.
+However, our `stub_deserialize` function assumes that this case will never happen!
+<!-- I'm not sure this case can be deemed an under-approximation just because of that. -->
+Therefore, the stub we wrote cannot be considered a **sound model** of `serde_json::from_slice`, and it's dangerous to use it in our harnesses.
+
+We hope this convinces you about the risks of stubbing.
+Always keep in mind that stubs are essentially additional assumptions in your harnesses.
+Because of that, we recommend our users to avoid stubbing whenever they can.
```

Comment:
Maybe turn the sentence positive: "we recommend our users to only use stubbing when necessary".